zcash_client_sqlite/src/wallet/scanning.rs

```diff
@@ -0,0 +1,1107 @@
+                    Joined::Two(left, right)
+                } else {
+                    Joined::One(left)
+                }
+            }
+            Dominance::Equal => Joined::One(ScanRange::from_parts(
+                left.block_range().start..max(left.block_range().end, right.block_range().end),
+                left.priority(),
+            )),
+            Dominance::Right => {
+                if let Some(left) = left.truncate_end(right.block_range().start) {
+                    if let Some(end) = left.truncate_start(right.block_range().end) {
+                        Joined::Three(left, right, end)
+                    } else {
+                        Joined::Two(left, right)
+                    }
+                } else if let Some(end) = left.truncate_start(right.block_range().end) {
+                    Joined::Two(right, end)
+                } else {
+                    Joined::One(right)
+                }
+            }
+        }
+    }
+
+    use RangeOrdering::*;
+    match RangeOrdering::cmp(to_insert.block_range(), current.block_range()) {
+        LeftFirstDisjoint => join_nonoverlapping(to_insert, current),
+        LeftFirstOverlap | RightContained => join_overlapping(to_insert, current, Insert::Left),
+        Equal => Joined::One(ScanRange::from_parts(
+            to_insert.block_range().clone(),
+            update_priority(current.priority(), to_insert.priority()),
+        )),
+        RightFirstOverlap | LeftContained => join_overlapping(current, to_insert, Insert::Right),
+        RightFirstDisjoint => join_nonoverlapping(current, to_insert),
+    }
+}
+
+#[derive(Debug, Clone)]
+enum SpanningTree {
+    Leaf(ScanRange),
+    Parent {
+        span: Range<BlockHeight>,
+        left: Box<SpanningTree>,
+        right: Box<SpanningTree>,
+    },
+}
+
+impl SpanningTree {
+    fn span(&self) -> Range<BlockHeight> {
+        match self {
+            SpanningTree::Leaf(entry) => entry.block_range().clone(),
+            SpanningTree::Parent { span, .. } => span.clone(),
+        }
+    }
+
+    fn from_joined(joined: Joined) -> Self {
+        match joined {
+            Joined::One(entry) => SpanningTree::Leaf(entry),
+            Joined::Two(left, right) => SpanningTree::Parent {
+                span: left.block_range().start..right.block_range().end,
+                left: Box::new(SpanningTree::Leaf(left)),
+                right: Box::new(SpanningTree::Leaf(right)),
+            },
+            Joined::Three(left, mid, right) => SpanningTree::Parent {
+                span: left.block_range().start..right.block_range().end,
+                left: Box::new(SpanningTree::Leaf(left)),
+                right: Box::new(SpanningTree::Parent {
+                    span: mid.block_range().start..right.block_range().end,
+                    left: Box::new(SpanningTree::Leaf(mid)),
+                    right: Box::new(SpanningTree::Leaf(right)),
+                }),
+            },
+        }
+    }
```

Comment:
Extracted the common logic from `SpanningTree::insert`:
```suggestion
    }

    fn from_insert(
        left: Box<Self>,
        right: Box<Self>,
        to_insert: ScanRange,
        insert: Insert,
    ) -> Self {
        let (left, right) = match insert {
            Insert::Left => (Box::new(left.insert(to_insert)), right),
            Insert::Right => (left, Box::new(right.insert(to_insert))),
        };
        SpanningTree::Parent {
            span: left.span().start..right.span().end,
            left,
            right,
        }
    }

    fn from_split(
        left: Box<Self>,
        right: Box<Self>,
        to_insert: ScanRange,
        split_point: BlockHeight,
    ) -> Self {
        let (l_insert, r_insert) = to_insert
            .split_at(split_point)
            .expect("Split point is within the range of to_insert");
        let left = Box::new(left.insert(l_insert));
        let right = Box::new(right.insert(r_insert));
        SpanningTree::Parent {
            span: left.span().start..right.span().end,
            left,
            right,
        }
    }
```